core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java

```diff
@@ -0,0 +1,393 @@
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.tasks.TaskManager;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrimaryReplicaSyncer extends AbstractComponent {
+
+    private final TaskManager taskManager;
+    private final SyncAction syncAction;
+
+    public static final ByteSizeValue DEFAULT_CHUNK_SIZE = new ByteSizeValue(512, ByteSizeUnit.KB);
+
+    private volatile ByteSizeValue chunkSize = DEFAULT_CHUNK_SIZE;
+
+    @Inject
+    public PrimaryReplicaSyncer(Settings settings, TransportService transportService, TransportResyncReplicationAction syncAction) {
+        this(settings, transportService.getTaskManager(), syncAction);
+    }
+
+    // for tests
+    public PrimaryReplicaSyncer(Settings settings, TaskManager taskManager, SyncAction syncAction) {
+        super(settings);
+        this.taskManager = taskManager;
+        this.syncAction = syncAction;
+    }
+
+    void setChunkSize(ByteSizeValue chunkSize) { // only settable for tests
+        if (chunkSize.bytesAsInt() <= 0) {
+            throw new IllegalArgumentException("chunkSize must be > 0");
+        }
+        this.chunkSize = chunkSize;
+    }
+
+    public void resync(IndexShard indexShard, ActionListener<ResyncTask> listener) throws IOException {
+        try (Translog.View view = indexShard.acquireTranslogView()) {
+            Translog.Snapshot snapshot = view.snapshot();
+            ShardId shardId = indexShard.shardId();
+
+            // Wrap translog snapshot to make it synchronized as it is accessed by different threads through SnapshotSender.
+            // Even though those calls are not concurrent, snapshot.next() uses non-synchronized state and is not multi-thread-compatible
+            // Also fail the resync early if the shard is shutting down
+            Translog.Snapshot wrappedSnapshot = new Translog.Snapshot() {
+
+                @Override
+                public synchronized int totalOperations() {
+                    return snapshot.totalOperations();
+                }
+
+                @Override
+                public synchronized Translog.Operation next() throws IOException {
+                    if (indexShard.state() != IndexShardState.STARTED) {
+                        throw new IndexShardNotStartedException(shardId, indexShard.state());
+                    }
+                    return snapshot.next();
+                }
+            };
+
+            resync(shardId, indexShard.routingEntry().allocationId().getId(), wrappedSnapshot,
+                indexShard.getGlobalCheckpoint() + 1, indexShard.seqNoStats().getMaxSeqNo(), listener);
```

Comment:
thinking about this more, why do we need the max setting here? we already acquired the snapshot, so this keeps us an implied maximum (as no ops will be added)?